modules/core/src/main/scala/Kernel.scala

```diff
@@ -4,9 +4,34 @@
 
 package natchez
 
+import cats.kernel.Monoid
+import natchez.Kernel.HeaderKey
+
 /**
- * An opaque hunk of data that can we can hand off to another system (in the form of HTTP headers),
- * which can then create new spans as children of this one. By this mechanism we allow our trace
- * to span remote calls.
- */
-final case class Kernel(toHeaders: Map[String, String])
+  * An opaque hunk of data that can we can hand off to another system (in the form of HTTP headers),
+  * which can then create new spans as children of this one. By this mechanism we allow our trace
+  * to span remote calls.
+  */
+final case class Kernel(keyedHeaders: Map[HeaderKey, String]) {
```

Comment:
We're assuming here that `HeaderKey` (an open trait) has sane universal equality and hashing, which I am reluctant to do. I wonder if it would work better to have `case class Provider(name: String)` and then `Map[Provider, Map[String, String]]`.

I think we may also want to prefix the provider to the header keys because we can still have the problem where two providers use the same key and it means different things (and one of them will be clobbered).

Reply:
Thanks for taking a look!

> We're assuming here that HeaderKey (an open trait) has sane universal equality and hashing, which I am reluctant to do

Ah yeah, I didn't think about this, bad times!

> I think we may also want to prefix the provider to the header keys because we can still have the problem where two providers use the same key and it means different things (and one of them will be clobbered).

I suppose we only want to do this when converting the keys from `Map[Provider, Map[String, String]]` to `Map[String, String]`, the various providers should be looking up their specific map in the map of maps.

Reply:
Right, each provider sees its own map. This still raises the possibility of concatenating two spans from the same provider, in which case the one on the left will win if there are key collisions. This is ok from a semigroup point of view but it will lead to puzzling behavior (although it will be hard to provoke)

Reply:
Ah yeah, unless we are able to assign some sort of identifer to the provider on creation which will ensure the headers are always mapped back correctly. Of course this only works when encoding headers and not decoding. The only way around that may be to have separate types for when continuing in an entry point and propagating context from a span.